File: tr_tool.py
from pathlib import Path
from check_file_permission import check_file_permission

def tr_file(file_path, from_chars, to_chars, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    new_content = content.translate(str.maketrans(from_chars, to_chars))
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(new_content)
    return f"Translated characters in {file_path}"

File: ts-tools/tr_tool.ts
import * as fs from 'fs';
import * as path from 'path';
import { checkFilePermission } from './check_file_permission';

export function trFile(
  filePath: string,
  fromChars: string,
  toChars: string,
  config: any,
): Promise<string> {
  return new Promise((resolve, reject) => {
    checkFilePermission(filePath, 'write', config)
      .then(() => {
        const resolvedPath = path.resolve(filePath);
        fs.readFile(resolvedPath, 'utf-8', (err, data) => {
          if (err) {
            reject(err);
          } else {
            const translationMap = new Map<string, string>();
            for (let i = 0; i < fromChars.length; i++) {
              translationMap.set(fromChars[i], toChars[i]);
            }
            let newData = '';
            for (const char of data) {
              newData += translationMap.get(char) || char;
            }
            fs.writeFile(resolvedPath, newData, 'utf-8', (err) => {
              if (err) {
                reject(err);
              } else {
                resolve(`Translated characters in ${filePath}`);
              }
            });
          }
        });
      })
      .catch(reject);
  });
}
